gh-51067: add `ZipFile.remove()`

danny0838 opened the pull request:
This is a revision of #19358 (for issue #51067) as the original author seems not keeping working.

Notable changes:

* Added docs and tests.

* Support mode 'w' and 'x', as noted by https://github.com/python/cpython/issues/51067#issuecomment-1093479250

* Support removing multiple members and removing non-physically with the internal `_remove_members` method, as they may be used by some interested people, as noted by https://github.com/python/cpython/issues/51067#issuecomment-1093479247 and https://github.com/python/cpython/issues/51067#issuecomment-1093479277.

  They are not currently introduced in the public `remove` API, as it would involve more complicated changes to the public APIs (e.g. introducing error handling for multiple members, and a extra method that purges stale data by non-physical removing) and other ZipFile related APIs do not support similar operations.

* Move physical data in chunks, to prevent a memory issue for large files.

* Fixed a flaw of the previous implementation that `self.NameToInfo` gets a missing key when removing one of duplicated arcnames.

---

<!-- gh-issue-number: gh-51067 -->
* Issue: gh-51067
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--103033.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

danny0838 committed:
Add `ZipFile.remove()` (#51067)

This is a revision of commit 659eb048cc9cac73c46349eb29845bc5cd630f09 (PR #19358), notably with following changes:

- Add documentation and tests.
- Raise `ValueError` for a bad mode, as in other methods.
- Support multi-member removal in `_remove_members()`.
- Support non-physical removal in `_remove_members()`.
- Move physical file data in chunks to prevent excessive memory usage on large files.
- Fix missing entry in `self.NameToInfo` when removing a duplicated archive name.
- Also update `ZipInfo._end_offset` for physically moved files.

Co-authored-by: Éric <merwok@netwok.org>

(cherry picked from commit e6bc82a3995f42df1c666c8889a2e4e4f938ff67 (PR #103033))

barneygale commented:
Also, please don't force-push to an open PR. It makes it harder for reviewers to follow changes! Thanks :)

merwok commented on Doc/library/zipfile.rst:
```diff
@@ -492,6 +492,23 @@ ZipFile Objects
+      Removing a member in an archive may involve a move of many internal data
+      records, which can be I/O intensive for a large ZIP file.
+
+   .. versionadded:: 3.12
```
```suggestion
   .. versionadded:: next
```

merwok commented on Lib/zipfile/__init__.py:
```diff
@@ -1691,6 +1691,32 @@ def extractall(self, path=None, members=None, pwd=None):
+            zinfo = zinfo_or_arcname
+            # make sure zinfo exists
+            if zinfo not in self.filelist:
+                raise KeyError(
+                    'There is no item %r in the archive' % zinfo_or_arcname)
```
```suggestion
                raise KeyError(f'There is no item {zinfo:r} in the archive')
```

merwok commented on Lib/zipfile/__init__.py:
```diff
@@ -1751,6 +1777,71 @@ def _extract_member(self, member, targetpath, pwd):
+        # get a sorted filelist by header offset, in case the dir order
+        # doesn't match the actual entry order
+        filelist = sorted(self.filelist, key=lambda x: x.header_offset)
+        for i in range(len(filelist)):
+            info = filelist[i]
```
```suggestion
        for i, info in enumerate(filelist):
```
no ❔

danny0838 committed:
Update `ZipInfo._end_offset`

danny0838 committed:
Fix and improve tests

danny0838 committed:
Test for zip64 more thoroughly

danny0838 committed:
No more allow mode 'w'/'x'

- File is not truncated in mode 'w'/'x', which results non-shrinked file.
- This cannot be simply resolved by adding truncation for mode 'w'/'x', which may be used on an unseekable file buffer and truncation is not allowed.

danny0838 committed:
Add tests for zstd

danny0838 committed:
Remove unneeded seek

- The seek will be automatically called in `ZipFile.close`.

danny0838 committed:
Add lock for thread safety

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

danny0838 committed:
Suppress duplicated name warning during generating test files

danny0838 committed:
Fix rst format

distantvale commented:
@danny0838 in reply to https://github.com/python/cpython/issues/51067#issuecomment-2985304746 (sorry about fragmenting the discussion) -

I'm generally supportive of the iterative approach of start simple and expand from that.

Having said that, since you already implemented `_remove_members`, I don't see why not expose it through a public API, given that ZipFile has other examples for functions that receive a list of members (`members` arg of `extractall`) - can simply follow the same pattern for consistency.
I do agree it's much more useful - my own local implementation only removes a single member and I do see the performance benefits of having an ability to remove multiple members, as many times I myself need to remove multiple members in many different files and it would be much beneficial.

I'm not fond of the idea of a 2 step removal (remove/repack), since it can leave the file in state that some may see (me included) as corruption - zip files aren't support to have a CD entry without a corresponding LF entry.
I'd rather take an approach that limits certain things (cannot remove a file from a self extracting archive for example),
and I don't see the large benefit of delayed repacking flexibility, seems like rather an edge case (but I might be wrong here).

Also some of the questions (how to handle a folder) are relevant regardless of whether we take the 1 step or 2 step approach.
If the 2 step approach has more support here, I'll take it - we can always improve it in the future, but I'd prefer to keep things simpler when possible.

And last but not least - perfect is the enemy of good - it's not the last python version, and we can always improve as long as the initial version is reasonable, which I believe it is.

merwok commented on Misc/NEWS.d/next/Core_and_Builtins/2025-05-22-12-52-35.gh-issue-51067.tJxGGF.rst:
```diff
@@ -0,0 +1 @@
+Add ``ZipFile.remove()``
```
```suggestion
Add :meth:`ZipFile.remove <zipfile.ZipFile.remove>`.
```

danny0838 committed:
Apply suggestion from @merwok

Co-authored-by: Éric <merwok@netwok.org>

gpshead commented:
#134627 implemented a version of this.
